language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -343,32 +190,83 @@ unliftConstant term = case term of
-unliftConstantEval
-    :: forall a m uni err.
-       ( MonadError (ErrorWithCause uni err) m, AsUnliftingError err
-       , GShow uni, GEq uni, uni `Includes` a
-       )
-    => DynamicBuiltinNameMeanings uni -> Evaluator Term uni m -> Term TyName Name uni () -> m a
-unliftConstantEval means eval = eval means >=> unliftConstant
-
--- Encode '()' from Haskell as @all r. r -> r@ from PLC.
--- This is a very special instance, because it's used to define functions that are needed for
--- other instances, so we keep it here.
-instance (GShow uni, GEq uni, uni `Includes` Integer) => KnownType uni () where
-    toTypeAst _ = unit
-
-    -- We need this matching, because otherwise Haskell expressions are thrown away rather than being
-    -- evaluated and we use 'unsafePerformIO' for logging, so we want to compute the '()' just
-    -- for side effects that the evaluation may cause.
-    makeKnown () = unitval
-
-    readKnown eval term = do
-        let integer = mkTyBuiltin @Integer ()
-            termApp = Apply () (TyInst () term integer) $ mkConstant @Integer () 1
-        i <- unliftConstantEval mempty eval termApp
-        if i == (1 :: Integer)
-            then pure ()
-            else throwingWithCause _UnliftingError "Not an integer-encoded ()" $ Just term
+{- Note [KnownType's defaults]
+We use @default@ for providing instances for built-in types instead of @DerivingVia@,
+because the latter breaks on
+
+    1. proxy a
+    2. m a
+
+It's possible to circumvent this, by using
+
+    1. Proxy a
+    2. forall r. (a -> m r) -> m r
+
+instead, but then we need to recover the original handy definitions and
+make the API and the code bloated (instances are more verbose with @DerivingVia@).
+-}
+
+-- See Note [KnownType's defaults]
+-- | Haskell types known to exist on the PLC side.
+class KnownType uni a where
+    -- | The type representing @a@ used on the PLC side.
+    toTypeAst :: proxy a -> Type TyName uni ()
+    default toTypeAst :: uni `Includes` a => proxy a -> Type TyName uni ()
+    toTypeAst _ = mkTyBuiltin @a ()
+
+    -- | Convert a Haskell value to the corresponding PLC term.
+    -- The inverse of 'readKnown'.
+    makeKnown :: a -> Term TyName Name uni ()
+    default makeKnown :: uni `Includes` a => a -> Term TyName Name uni ()
+    -- We need @($!)@, because otherwise Haskell expressions are thrown away rather than being
+    -- evaluated and we use 'unsafePerformIO' for logging, so we want to compute the Haskell value
+    -- just for side effects that the evaluation may cause.
+    makeKnown x = mkConstant () $! x
+
+    -- | Convert a PLC term to the corresponding Haskell value.
+    -- The inverse of 'makeKnown'.
+    readKnown
+        :: (MonadError (ErrorWithCause uni err) m, AsUnliftingError err)
+        => Term TyName Name uni () -> m a
+    default readKnown
+        :: ( MonadError (ErrorWithCause uni err) m, AsUnliftingError err
+           , GShow uni, GEq uni, uni `Includes` a
```

Comment:
This is `KnownBuiltinType` from below?

Reply:
Maybe all the default methods should have that constraint - it's too much for some, but it indicates the intended default (it's for builtins).